Gems/SceneProcessing/Code/Source/Generation/Components/MeshOptimizer/MeshOptimizerComponent.cpp

```diff
@@ -261,6 +264,45 @@ namespace AZ::SceneGenerationComponents
         ).empty();
     }
 
+    static AZ::Outcome<SceneData::GraphData::CustomPropertyData::PropertyMap*> FindOrCreateCustomPropertyData(
+        Containers::SceneGraph& graph, const Containers::SceneGraph::NodeIndex& nodeIndex)
+    {
+        NodeIndex customPropertyIndex =
+            SceneAPI::Utilities::GetImmediateChildOfType(graph, nodeIndex, azrtti_typeid<SceneAPI::DataTypes::ICustomPropertyData>());
+
+        if (!customPropertyIndex.IsValid())
+        {
+            // If no custom property data node exists, insert one
+            AZStd::shared_ptr<SceneData::GraphData::CustomPropertyData> createdCustumPropertyData =
+                AZStd::make_shared<SceneData::GraphData::CustomPropertyData>();
+            customPropertyIndex = graph.AddChild(nodeIndex, "custom_properties", AZStd::move(createdCustumPropertyData));
+        }
+
+        SceneData::GraphData::CustomPropertyData* customPropertyDataNode =
+            azrtti_cast<SceneData::GraphData::CustomPropertyData*>(graph.GetNodeContent(customPropertyIndex).get());
+
+        // Now look up the optimized index
+        SceneData::GraphData::CustomPropertyData::PropertyMap& propertyMap = customPropertyDataNode->GetPropertyMap();
+
+        if (auto iter = propertyMap.find(SceneAPI::Utilities::OptimizedMeshPropertyMapKey); iter != propertyMap.end())
+        {
+            const auto& [key, anyNodeIndex] = *iter;
+            if (!anyNodeIndex.empty() && anyNodeIndex.is<NodeIndex>())
+            {
+                NodeIndex optimizedMeshIndex = AZStd::any_cast<NodeIndex>(anyNodeIndex);
+                AZ_TracePrintf(
+                    AZ::SceneAPI::Utilities::LogWindow,
+                    "Optimized mesh already exists at '%s', there must be multiple mesh groups that have selected this mesh. Skipping "
+                    "the additional ones.",
+                    graph.GetNodeName(optimizedMeshIndex).GetName());
+
+                return AZ::Failure();
+            }
```

Comment:
[suggestion] IMO returning a failure here is a bit odd in a function that's called `FindOrCreate...`, it'd seem like if it found one already then it should return success too. My suggestion would be to return something similar to `map::insert`, which returns instead a `pair<PropertyMap*,bool>` where the bool can be used by the caller to indicate if it's an existing mesh or not. And the first element of the pointer can be used for the case where there is an actual error because the pointer can be `nullptr`.

Reply:
With my latest changes, I ended up calling this function twice, and one time I needed to look up if the optimized node already exists, and the other didn't. So I split this into two functions, which makes more sense anyways since looking for an existing optimized mesh goes beyond the purview of a function named "FindOrCreateCustomPropertyData.